lib/api/BackbeatServer.js

```diff
@@ -0,0 +1,233 @@
+                    `path ${backbeatRequest.getRoute()}`,
+                code: 400,
+            });
+            return this._errorResponse(backbeatRequest);
+        }
+
+        return true;
+    }
+
+    /**
+     * Check if Kafka and Zookeeper are working properly
+     * @param {BackbeatRequest} backbeatRequest - Backbeat request object
+     * @return {boolean} true if no errors
+     */
+    _areConditionsOk(backbeatRequest) {
+        if (!this.backbeatAPI.isConnected()) {
+            backbeatRequest.setError({
+                error: 'Internal Error',
+                code: 500,
+            });
+            return this._errorResponse(backbeatRequest);
+        }
+
+        return true;
+        // Returns 500 when Kafka is unreachable
+        // Returns 500 when Zookeeper is unreachable
+        // Returns 500 if any node has fewer ISRs than normal
+    }
+
+    /**
+     * Server's error response handler
+     * @param {BackbeatRequest} backbeatRequest - Backbeat request object
+     * @return {undefined}
+     */
+    _errorResponse(backbeatRequest) {
+        this._response(backbeatRequest, backbeatRequest.getError());
+    }
+
+    /**
+     * Server incoming request handler
+     * @param {object} req - request object
+     * @param {object} res - response object
+     * @return {undefined}
+     */
+    _requestListener(req, res) {
+        req.socket.setNoDelay();
+        const bbRequest = new BackbeatRequest()
+            .setRequest(req)
+            .setResponse(res)
+            .setLog(this._logger.newRequestLogger())
+            .setRoute(req.url);
+
+        // check request conditions and all 3rd-party conditions here
+        // should ISRs always be checked as well?
+        // What to name `_areConditionsOk`?
+        if (this._isValidRequest(req, bbRequest)
+        && this._areConditionsOk(bbRequest)) {
+            bbRequest.setStatusCode(200);
+            bbRequest.setRoute(bbRequest.getRoute().slice(3));
+
+            this.backbeatAPI[bbRequest.getRoute()](bbRequest, this._response);
+        }
+    }
+
+    /**
+     * Server's response to the client
+     * @param {BackbeatRequest} backbeatRequest - Backbeat request object
+     * @param {object} data - response to send to client
+     * @return {object} res - response object
+     */
+    _response(backbeatRequest, data) {
+        const log = backbeatRequest.getLog();
+        const req = backbeatRequest.getRequest();
+        const res = backbeatRequest.getResponse();
+        log.trace('writing HTTP response', {
+            method: 'BackbeatServer._response',
+        });
+        const code = backbeatRequest.getStatusCode();
+        const payload = Buffer.from(JSON.stringify(data), 'utf8');
+
+        res.writeHead(code, {
+            'content-type': 'application/json',
+            'content-length': payload.length,
+        });
+        res.write(payload);
+        BackbeatServer.logRequestEnd(log, req, res);
+        return res.end();
```

Comment:
minor: you can avoid the `res.write` by providing the contents to `res.end` directly.